sdk/src/azure/platform/az_mosquitto.c

```diff
@@ -0,0 +1,278 @@
+  return (az_mqtt_options){ .certificate_authority_trusted_roots = NULL,
+                            .openssl_engine = NULL,
+                            .mosquitto_handle = NULL };
+}
+
+AZ_NODISCARD az_result az_mqtt_init(az_mqtt* mqtt, az_mqtt_options const* options)
+{
+  _az_PRECONDITION_NOT_NULL(mqtt);
+  mqtt->_internal.options = options == NULL ? az_mqtt_options_default() : *options;
+  mqtt->mosquitto_handle = mqtt->_internal.options.mosquitto_handle;
+  mqtt->_internal.platform_mqtt.pipeline = NULL;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+az_mqtt_outbound_connect(az_mqtt* mqtt, az_context* context, az_mqtt_connect_data* connect_data)
+{
+  (void)context;
+  az_result ret = AZ_OK;
+  az_mqtt* me = (az_mqtt*)mqtt;
+
+  // IMPORTANT: application must call mosquitto_lib_init() before any Mosquitto clients are created.
+
+  if (me->mosquitto_handle == NULL)
+  {
+    me->mosquitto_handle = mosquitto_new(
+        (char*)az_span_ptr(connect_data->client_id),
+        false, // clean-session
+        me); // callback context - i.e. user data.
+  }
+
+  if (me->mosquitto_handle == NULL)
+  {
+    ret = AZ_ERROR_OUT_OF_MEMORY;
+    return ret;
+  }
+
+  /* Configure callbacks. This should be done before connecting ideally. */
+  mosquitto_log_callback_set(me->mosquitto_handle, _az_mosquitto_on_log);
+  mosquitto_connect_callback_set(me->mosquitto_handle, _az_mosquitto_on_connect);
+  mosquitto_disconnect_callback_set(me->mosquitto_handle, _az_mosquitto_on_disconnect);
+  mosquitto_publish_callback_set(me->mosquitto_handle, _az_mosquitto_on_publish);
+  mosquitto_subscribe_callback_set(me->mosquitto_handle, _az_mosquitto_on_subscribe);
+  mosquitto_unsubscribe_callback_set(me->mosquitto_handle, _az_mosquitto_on_unsubscribe);
+  mosquitto_message_callback_set(me->mosquitto_handle, _az_mosquitto_on_message);
+
+  if (az_span_size(me->_internal.options.certificate_authority_trusted_roots) != 0)
+  {
+    _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_tls_set(
+        me->mosquitto_handle,
+        (const char*)az_span_ptr(me->_internal.options.certificate_authority_trusted_roots),
+        NULL,
+        (const char*)az_span_ptr(connect_data->certificate.cert),
+        (const char*)az_span_ptr(connect_data->certificate.key),
+        NULL)));
+  }
+
+  _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_username_pw_set(
+      me->mosquitto_handle,
+      (const char*)az_span_ptr(connect_data->username),
+      (const char*)az_span_ptr(connect_data->password))));
+
+  _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_connect_async(
+      (struct mosquitto*)me->mosquitto_handle,
+      (char*)az_span_ptr(connect_data->host),
+      connect_data->port,
+      AZ_MQTT_KEEPALIVE_SECONDS)));
+
+  _az_RETURN_IF_FAILED(_az_result_from_mosq(mosquitto_loop_start(me->mosquitto_handle)));
+
+  return ret;
+}
+
+AZ_NODISCARD az_result
+az_mqtt_outbound_sub(az_mqtt* mqtt, az_context* context, az_mqtt_sub_data* sub_data)
+{
+  (void)context;
```

Comment:
Please apply context checking prior to all operations. 
If the application cancelled the operation, we should not call the I/O function.

See `az_context_has_expired()` - will require getting the current time.

Reply:
Removed `az_context` from `az_mqtt.h` moved to `az_mqtt_policy` instead